python/ray/tune/trial_executor.py

```diff
@@ -0,0 +1,163 @@
+                            stop_logger=False)
+            trial.result_logger.flush()
+            self.start_trial(trial)
+        except Exception:
+            error_msg = traceback.format_exc()
+            print("Error recovering trial from checkpoint, abort:", error_msg)
+            self.stop_trial(trial, error=True, error_msg=error_msg)
+
+    def on_scheduler_decision(self, trial, decision):
+        """A hook called when TrialScheduler make a decision.
+
+        Args:
+            trial (Trial): the scheduler decide whether to continue this trial.
+            decision (str): (CONTINUE,PAUSE,STOP) constant from TrialScheduler.
+        """
+        pass
+
+    def pause_trial(self, trial):
+        """Pauses the trial.
+
+        We want to release resources (specifically GPUs) when pausing an
+        experiment. This results in PAUSED state that similar to TERMINATED.
+        """
+        assert trial.status == Trial.RUNNING, trial.status
+        try:
+            self.save(trial, Checkpoint.MEMORY)
+            self.stop_trial(trial, stop_logger=False)
+            trial.status = Trial.PAUSED
+        except Exception:
+            print("Error pausing runner:", traceback.format_exc())
+            trial.status = Trial.ERROR
+
+    def unpause_trial(self, trial):
+        """Sets PAUSED trial to pending to allow scheduler to start."""
+        assert trial.status == Trial.PAUSED, trial.status
+        trial.status = Trial.PENDING
+
+    def resume_trial(self, trial):
+        """Resumes PAUSED trials. This is a blocking call."""
+        assert trial.status == Trial.PAUSED, trial.status
+        self.start_trial(trial)
+
+    def get_running_trials(self):
+        """Returns all running trials."""
+        raise NotImplementedError("Subclasses of TrialExecutor must provide "
+                                  "get_running_trials() method")
+
+    def on_step_begin(self):
+        """A hook called before running one step of the trial event loop."""
+        pass
+
+    def on_step_end(self):
+        """A hook called after running one step of the trial event loop."""
+        pass
+
+    def fetch_one_result(self):
+        """Fetches one result from running trials.
+
+        It's a blocking call waits until one result is ready.
+
+        Return:
+            a tuple of (trial, result). if fetch result failed,
+            return (trial, None) other than raise Exception.
+        """
+        raise NotImplementedError("Subclasses of TrialExecutor must provide "
+                                  "fetch_one_result() method")
+
+    def debug_string(self):
+        """Returns a human readable message for printing to the console."""
+        pass
+
+    def restore(self, trial, checkpoint=None):
+        """Restore the state of trial from checkpoint or trial._checkpoint.
+
+        If restoring fails, the trial status will be set to ERROR.
+
+        Args:
+            trial (Trial): trial to be restored.
```

Comment:
Capitalize after colon - "Trial to be restored."

This shows up multiple times; please make sure this is consistent. Thanks!